rust/src/interpreter/code_state.rs

```diff
@@ -0,0 +1,265 @@
+use std::{cmp::min, mem, ops::Deref};
+
+use evmc_vm::StatusCode;
+
+use crate::types::{code_byte_type, u256, CodeByteType, Opcode};
+
+#[derive(Debug)]
+pub struct CodeState<'a> {
+    code: &'a [u8],
+    code_byte_type: Box<[CodeByteType]>,
+    pc: usize,
+}
+
+impl<'a> Deref for CodeState<'a> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.code
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum GetOpcodeError {
+    OutOfRange,
+    Invalid,
+}
+
+impl<'a> CodeState<'a> {
+    pub fn new(code: &'a [u8], pc: usize) -> Self {
+        Self {
+            code,
+            code_byte_type: compute_code_byte_types(code),
+            pc,
+        }
+    }
+
+    pub fn get(&self) -> Result<Opcode, GetOpcodeError> {
+        if self.pc >= self.code.len() {
+            Err(GetOpcodeError::OutOfRange)
+        } else if self.code_byte_type[self.pc] == CodeByteType::DataOrInvalid {
+            Err(GetOpcodeError::Invalid)
+        } else {
+            let op = self.code[self.pc];
+            let op = unsafe {
+                // SAFETY:
+                // [Opcode] has repr(u8) and therefore the same memory layout as u8.
+                // In get_code_byte_types this byte of the code was determined to be a valid opcode.
+                // Therefore the value is a valid [Opcode].
+                mem::transmute::<u8, Opcode>(op)
```

Comment:
Curiosity Question: is this for efficiency because it is faster than implementing a `From<u8>` on the Opcode or is it mainly for simplicity to not have to implement the `From<u8>`?

I guess the "unsafe" part of it is if there is a `u8` value read for which there would actually not be an Opcode enum value ...

Reply:
There is no safe way to convert `u8` to `Opcode` except for matching for each opcode again. However this check was already done when doing the jump analysis.

Reply:
I guess the question was more about implementing `From<u8>`
I guess the decision of not doing so is that such a from function will always succeed, even with invalid input, leading to undefined behaviour. 
On the other hand, this specific location seems to assume that this is a valid opcode value, and therefore forces the cast. 

Another alternative would be to implement TryFrom<u8> which would return an error if not possible. Such implementation would have the overhead of testing all valid codes each time. 

I think it is rather verbose, but unsafe block may be the right choice.